ports/rp2/machine_pin.c

```diff
@@ -261,155 +223,101 @@ static const mp_arg_t allowed_args[] = {
-            if (self->is_output) {
-                // todo need to disable output
-            }
-            self->is_output = false;
-        } else if (mode == GPIO_MODE_OUT) {
-            if (!self->is_output) {
-                // todo need to enable output
-                // for now we just set the value
-                if (value == -1) {
-                    value = self->last_output_value;
-                }
-                self->last_output_value = !self->last_output_value; // defeat shortcircuit
-                update_cyw43_value(self, value);
-                self->is_output = true;
-            }
-        } else {
-            mp_raise_ValueError("only Pin.OUT and Pin.IN are supported for this pin");
-        }
-    }
-
-    if (value != -1) {
-        if (self->is_output) {
-            update_cyw43_value(self, value);
-        } else {
-            // figure if you pass a value to IN it should still remember it (this is what regular GPIO does)
-            self->last_output_value = value;
+        // set initial value (do this before configuring mode/pull)
+        if (!is_ext_pin(self)) {
+            gpio_put(self->id, value);
         }
     }
 
-    return mp_const_none;
-}
-#endif
-
-STATIC mp_obj_t machine_pin_obj_init_helper(const machine_pin_obj_t *self, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
-
-    // parse args
-    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
-    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
-
-    // set initial value (do this before configuring mode/pull)
-    if (args[ARG_value].u_obj != mp_const_none) {
-        gpio_put(self->id, mp_obj_is_true(args[ARG_value].u_obj));
-    }
-
     // configure mode
     if (args[ARG_mode].u_obj != mp_const_none) {
         mp_int_t mode = mp_obj_get_int(args[ARG_mode].u_obj);
-        if (mode == GPIO_MODE_IN) {
+        if (is_ext_pin(self)) {
+            #if MICROPY_HW_PIN_EXT_COUNT
+            // The regular Pins are const, but the external pins are mutable.
+            machine_pin_obj_t *mutable_self = (machine_pin_obj_t *)self;
+            machine_pin_ext_config(mutable_self, mode, value);
+            #endif
+        } else if (mode == MACHINE_PIN_MODE_IN) {
             mp_hal_pin_input(self->id);
-        } else if (mode == GPIO_MODE_OUT) {
+        } else if (mode == MACHINE_PIN_MODE_OUT) {
             mp_hal_pin_output(self->id);
-        } else if (mode == GPIO_MODE_OPEN_DRAIN) {
+        } else if (mode == MACHINE_PIN_MODE_OPEN_DRAIN) {
             mp_hal_pin_open_drain(self->id);
         } else {
-            // Alternate function.
-            gpio_set_function(self->id, args[ARG_alt].u_int);
+            // Configure alternate function.
+            mp_uint_t af = args[ARG_alt].u_int;
+            if (machine_pin_find_af(self, af) == NULL) {
+                mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("Invalid pin af: %d."), af);
```

Comment:
Please rename all uses of `af` to `alt` (eg `machine_pin_find_alt()`).

Also error messages should start with a lowercase letter (unless it's a class name).

Reply:
Renamed and fixed error messages.